File: frontend/lib/types/pad.ts
/** Contracts for the Visit Pad (mirrors backend/app/schemas/pad_schemas.py). */
import type { Department } from "@/lib/types/core";

export type SectionKind =
  | "text"
  | "fields"
  | "list"
  | "ai"
  /** Prescribed lines. Signing the pad issues the prescription from them. */
  | "medicines"
  /** Advised tests. Signing the pad places the order for the coded ones. */
  | "investigations";
export type FieldType =
  | "text"
  | "textarea"
  | "number"
  | "date"
  | "select"
  | "multiselect"
  | "checkbox";
export type AISource =
  | "intake_summary"
  | "red_flags"
  | "differentials"
  | "suggested_investigations";

export interface FieldSpec {
  key: string;
  label: string;
  type: FieldType;
  unit?: string | null;
  options: string[];
  required: boolean;
  catalogue_category?: string | null;
}

export interface SectionSpec {
  key: string;
  title: string;
  kind: SectionKind;
  visible_in_pad: boolean;
  visible_in_print: boolean;
  carry_forward: boolean;
  catalogue_category?: string | null;
  placeholder?: string | null;
  fields: FieldSpec[];
  ai_source?: AISource | null;
}

export type FieldValue = string | number | boolean | string[] | null;

/** One prescribed line as the pad holds it, in the prescription's own shape. */
export interface PadMedicine {
  name: string;
  formulary_code?: string | null;
  generic?: string | null;
  form?: string | null;
  strength?: string | null;
  dosage?: string | null;
  frequency_code?: string | null;
  frequency_text?: string | null;
  duration?: string | null;
  timing?: string | null;
  route?: string | null;
  instructions?: string | null;
  /** How the line arrived: read off a dictation, picked, or typed. */
  source?: "dictated" | "catalog" | "manual";
}

/** One advised test. Without a code it prints as advice but places no order. */
export interface PadInvestigation {
  code?: string | null;
  name: string;
  note?: string | null;
}

/** One section's content. Which keys are present depends on the kind. */
export interface SectionValue {
  text?: string;
  items?: string[];
  fields?: Record<string, FieldValue>;
  /** What the doctor has accepted. Only these are prescribed or ordered. */
  medicines?: PadMedicine[];
  investigations?: PadInvestigation[];
  /**
   * Waiting to be tapped across — heard in the dictation, or advised by the
   * intake. Nothing here is prescribed, ordered or printed. Plain strings on
   * a list section, rows on a medicines or investigations one.
   */
  suggestions?: string[] | PadMedicine[] | PadInvestigation[];
}

export interface SectionOrigin {
  source: string;
  drafted_at?: string;
  edited?: boolean;
  edited_by?: string;
  template_id?: string;
  name?: string;
  document_id?: string;
  signed_at?: string | null;
}

export type PadStatus = "draft" | "signed" | "superseded";

export type DocumentFamily = "certificate" | "consent";

export interface PadDocumentTypeInfo {
  key: string;
  label: string;
  scope: string;
  authority: "doctor" | "nursing";
  family?: DocumentFamily | null;
  /** A link the document cannot be written without, e.g. "admission". */
  requires?: string | null;
}

/** The fixed wording a consent form prints. */
export interface FormWording {
  key: string;
  label: string;
  label_hi?: string | null;
  family: DocumentFamily;
  requires?: string | null;
  statement?: { en: string[]; hi: string[] } | null;
}

export interface PadDocument {
  id: string;
  document_type: string;
  title: string;
  patient_id: string;
  /** The prescription this pad issued when it was signed. */
  prescription_id?: string | null;
  investigation_order_id?: string | null;
  consultation_id?: string | null;
  admission_id?: string | null;
  surgery_id?: string | null;
  department?: Department | null;
  layout_revision: number;
  sections: SectionSpec[];
  values: Record<string, SectionValue>;
  provenance: Record<string, SectionOrigin>;
  status: PadStatus;
  author_name: string;
  signed_by_name?: string | null;
  signed_at?: string | null;
  group_id: string;
  version: number;
  supersedes_id?: string | null;
  amendment_reason?: string | null;
  print_count: number;
  /** Certificates and consent forms: given at first signing, kept by corrections. */
  serial_number?: string | null;
  paper_signed_at?: string | null;
  paper_signed_by_name?: string | null;
  created_at: string;
  updated_at: string;
}

export interface PadDocumentSummary {
  id: string;
  document_type: string;
  title: string;
  consultation_id?: string | null;
  admission_id?: string | null;
  surgery_id?: string | null;
  serial_number?: string | null;
  paper_signed_at?: string | null;
  status: PadStatus;
  version: number;
  author_name: string;
  signed_by_name?: string | null;
  signed_at?: string | null;
  created_at: string;
}

export interface PadTemplate {
  id: string;
  document_type: string;
  name: string;
  description?: string | null;
  owner_name: string;
  shared: boolean;
  use_count: number;
  section_keys: string[];
}

export type LayoutScope = "personal" | "department" | "hospital";

export interface PadLayout {
  document_type: string;
  document_label: string;
  scope: LayoutScope | "default";
  name: string;
  sections: SectionSpec[];
  revision: number;
  updated_by_name?: string | null;
  variables: Record<string, string>;
  /** Section key -> field keys the system needs; they cannot be removed. */
  protected?: Record<string, string[]>;
  /** Fixed layout: certificates, consents, radiology. */
  locked?: boolean;
}

export interface CatalogueSuggestion {
  text: string;
  use_count: number;
}

export interface ArrangementItem {
  key: string;
  visible_in_pad: boolean;
  visible_in_print: boolean;
}

File: backend/app/models/pad.py
"""The Visit Pad: configurable clinical documents.

Every clinical screen in the old system — the OPD pad, the nursing initial
assessment, the progress note, the discharge summary, the OT note, a medical
certificate — was the same engine wearing a different layout. This is that
engine. Four tables, each holding a different kind of thing:

  PadLayout      what a document looks like: its sections, in order, and for
                 each one whether it shows on screen, whether it prints, and
                 whether it carries forward to the next visit
  PadDocument    one filled document for one patient
  PadTemplate    a saved set of values a doctor loads instead of typing
  PadCatalogue   the suggestion list that grows as staff type

`document_type` is a plain string rather than a database enum on purpose.
Part 3 adds inpatient forms, theatre notes and certificates on top of this
engine, and a Postgres enum can gain values but never lose or rename them. A
document type is configuration, and it should be as cheap to add as a row.

A signed document is never edited. A correction is a new version in the same
group, and the old one is marked superseded — the same rule inpatient notes
already follow, for the same reason: the record of what was believed at the
time is the thing a clinical document exists to preserve.
"""
import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional

from sqlalchemy import (
    DateTime,
    Enum as SAEnum,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, TimestampMixin, UUIDPrimaryKeyMixin
from app.models.enums import Department, PadStatus

_VALUES = lambda e: [m.value for m in e]  # noqa: E731  (see models/investigation.py)


class PadLayout(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    """The shape of one kind of document, for one scope.

    Scopes, from most to least specific: a doctor's own layout within a
    department, the department's layout, the hospital's layout. A document
    type nobody has configured falls back to the built-in default in
    `app/pads/defaults.py`, so a new form is usable before anyone edits it.
    """

    __tablename__ = "pad_layouts"
    __table_args__ = (
        Index("ix_pad_layouts_scope", "document_type", "department", "owner_id"),
    )

    document_type: Mapped[str] = mapped_column(String(48), nullable=False, index=True)
    department: Mapped[Optional[Department]] = mapped_column(
        SAEnum(Department, name="department", values_callable=_VALUES, create_type=False)
    )
    # NULL means the layout belongs to the department or the hospital. A
    # personal layout lets one surgeon keep the order they think in without
    # imposing it on every colleague.
    owner_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), index=True
    )
    name: Mapped[str] = mapped_column(String(160), nullable=False)
    sections: Mapped[List[Dict[str, Any]]] = mapped_column(JSONB, nullable=False, default=list)
    # Incremented on every save, and copied onto each document created from
    # it, so a document can say which revision of its layout it was made on.
    revision: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    updated_by_name: Mapped[str] = mapped_column(String(255), nullable=False, default="")


class PadDocument(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    """One filled clinical document."""

    __tablename__ = "pad_documents"
    __table_args__ = (
        Index("ix_pad_documents_patient_type", "patient_id", "document_type", "created_at"),
        Index("ix_pad_documents_group", "group_id", "version"),
    )

    document_type: Mapped[str] = mapped_column(String(48), nullable=False, index=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False, default="")

    patient_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("patients.id", ondelete="CASCADE"),
        nullable=False, index=True,
    )
    consultation_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("consultations.id", ondelete="SET NULL"), index=True
    )
    admission_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("admissions.id", ondelete="SET NULL"), index=True
    )
    # Theatre notes belong to a surgery. They also carry the admission, when
    # there is one, so the case sheet and the records bundle find them.
    surgery_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("surgeries.id", ondelete="SET NULL"), index=True
    )
    department: Mapped[Optional[Department]] = mapped_column(
        SAEnum(Department, name="department", values_callable=_VALUES, create_type=False)
    )

    # The layout as it stood when this document was made. Printing a
    # two-year-old discharge summary must produce what was signed, not what
    # the layout says today after somebody reordered its sections.
    layout_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("pad_layouts.id", ondelete="SET NULL")
    )
    layout_revision: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    sections: Mapped[List[Dict[str, Any]]] = mapped_column(JSONB, nullable=False, default=list)

    # Section key -> that section's content. The shape depends on the kind of
    # section and is validated in `app/pads/sections.py`.
    values: Mapped[Dict[str, Any]] = mapped_column(JSONB, nullable=False, default=dict)

    # Section key -> where its first draft came from. An AI-drafted section
    # stays marked as such after a doctor edits it, because the reader of a
    # signed document is entitled to know what wrote the first version.
    provenance: Mapped[Dict[str, Any]] = mapped_column(JSONB, nullable=False, default=dict)

    status: Mapped[PadStatus] = mapped_column(
        SAEnum(PadStatus, name="pad_status", values_callable=_VALUES),
        nullable=False, default=PadStatus.DRAFT, index=True,
    )

    author_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL")
    )
    author_name: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    signed_by_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL")
    )
    signed_by_name: Mapped[Optional[str]] = mapped_column(String(255))
    signed_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))

    # --- version chain ------------------------------------------------------
    group_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False, default=uuid.uuid4)
    version: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    supersedes_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("pad_documents.id", ondelete="SET NULL")
    )
    amendment_reason: Mapped[Optional[str]] = mapped_column(Text)

    # Every print after the first is a duplicate, and says so on the page.
    print_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    # Certificates and consent forms only. The serial is given at first
    # signing and kept by corrections; the paper columns record that the
    # patient's signed copy of a consent form came back.
    # What signing this document issued. Filled once, when it is signed.
    prescription_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("prescriptions.id", ondelete="SET NULL")
    )
    investigation_order_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("investigation_orders.id", ondelete="SET NULL")
    )

    serial_number: Mapped[Optional[str]] = mapped_column(String(24), index=True)
    #: A radiology report's imaging order item.
    order_item_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("investigation_order_items.id", ondelete="SET NULL"), index=True
    )
    paper_signed_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    paper_signed_by_name: Mapped[Optional[str]] = mapped_column(String(255))


class PadTemplate(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    """A saved set of section values, loaded into a document instead of typed.

    Distinct from copying the previous visit: a template is the same content
    for many patients ("post-op knee advice"), a previous visit is one
    patient's own history.
    """

    __tablename__ = "pad_templates"

    document_type: Mapped[str] = mapped_column(String(48), nullable=False, index=True)
    department: Mapped[Optional[Department]] = mapped_column(
        SAEnum(Department, name="department", values_callable=_VALUES, create_type=False)
    )
    # NULL means shared with the department.
    owner_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), index=True
    )
    owner_name: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    name: Mapped[str] = mapped_column(String(160), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(Text)
    values: Mapped[Dict[str, Any]] = mapped_column(JSONB, nullable=False, default=dict)
    use_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    last_used_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))


class PadCatalogueEntry(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    """One phrase staff have typed before, in one category.

    The single biggest reason the old system felt fast. Nobody curates it: a
    diagnosis typed and saved becomes a suggestion the next time anyone
    starts typing in a diagnosis field, and the phrases used most rise to the
    top. A hospital's vocabulary is its own — "B/L OA knee", not the textbook
    — and this is where it accumulates.

    `department` is a plain string with "" for hospital-wide, not a nullable
    enum, because a unique constraint treats every NULL as distinct and would
    let the same phrase be stored a hundred times.
    """

    __tablename__ = "pad_catalogue"
    __table_args__ = (
        UniqueConstraint("category", "department", "normalized", name="uq_pad_catalogue_phrase"),
        Index("ix_pad_catalogue_lookup", "category", "department", "use_count"),
    )

    category: Mapped[str] = mapped_column(String(64), nullable=False)
    department: Mapped[str] = mapped_column(String(32), nullable=False, default="")
    text: Mapped[str] = mapped_column(Text, nullable=False)
    # Lower-cased, whitespace-collapsed. "B/L OA  Knee" and "b/l oa knee" are
    # one phrase used twice, not two phrases used once.
    normalized: Mapped[str] = mapped_column(String(255), nullable=False)
    use_count: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    last_used_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
